File: Array/56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Stack<int[]> stk = new Stack<>();

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        for(int[] interval: intervals) {

            if(!stk.isEmpty() && stk.peek()[1] >= interval[0]) {
                int[] peek = stk.pop();
                peek[1] = Math.max(interval[1], peek[1]);
                stk.push(peek);
            }
            else {
                stk.push(interval);
            }
        }

        int[][] out = new int[stk.size()][2];
        int i = stk.size()-1;
        while(!stk.isEmpty()){
            out[i--] = stk.pop();
        }

        return out;
    }
}

File: Array/56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        stk = []
        intervals.sort()

        for st, end in intervals:
            if stk and stk[-1][1] >= st:
                top = stk.pop()
                stk.append([top[0], max(end, top[1])])
            else:
                stk.append([st, end])
        return stk